src/main/java/com/codejoust/main/service/GameManagementService.java

```diff
@@ -283,14 +307,131 @@ public GameDto manuallyEndGame(String roomId, EndGameRequest request) {
+        GameReport gameReport = new GameReport();
+        int numProblems = game.getProblems().size();
+        int numPlayers = game.getPlayers().size();
+
+        // Initialize the statistic variables for each problem.
+        int[] userSolved = new int[numProblems];
+        double[] totalTestCasesPassed = new double[numProblems];
+        double[] totalAttemptCount = new double[numProblems];
+        for (Player player : game.getPlayers().values()) {
+            // For each user, add the relevant submission info.
+            User user = player.getUser();
+
+            // Construct the new submission group report for each user.
+            SubmissionGroupReport submissionGroupReport = new SubmissionGroupReport();
+            submissionGroupReport.setGameReportId(gameReport.getGameReportId());
+
+            // Iterate through each submission and update group statistics.
+            boolean[] problemsSolved = new boolean[numProblems];
+            int[] testCasesPassed = new int[numProblems];
+            for (Submission submission : player.getSubmissions()) {
+                int problemIndex = submission.getProblemIndex();
+                totalAttemptCount[problemIndex]++;
+
+                // If the problem was solved, set boolean value to true.
+                if (submission.getNumTestCases() == submission.getNumCorrect() && !problemsSolved[problemIndex]) {
+                    problemsSolved[problemIndex] = true;
+                    userSolved[problemIndex]++;
+                }
+
+                // Get the maximum number of test cases passed for each problem.
+                testCasesPassed[problemIndex] = Math.max(testCasesPassed[problemIndex], submission.getNumCorrect());
+
+                submissionGroupReport.addSubmissionReport(SubmissionMapper.toSubmissionReport(submission));
+            }
+
+            // Iterate through the test cases passed to add to the game total.
+            for (int i = 0; i < testCasesPassed.length; i++) {
+                totalTestCasesPassed[i] += testCasesPassed[i];
+            }
+            
+            // Set the problems and test cases statistics.
+            submissionGroupReport.setProblemsSolved(compactProblemsSolved(problemsSolved));
+            submissionGroupReport.setNumTestCasesPassed(Arrays.stream(testCasesPassed).sum());
+            
+            // Add the submission group report and the user.
+            user.addSubmissionGroupReport(submissionGroupReport);
+            gameReport.addUser(user);
+        }
+
+        // Set problem container variables.
+        int numTestCases = 0;
+        List<Problem> problems = game.getProblems();
+        for (int i = 0; i < numProblems; i++) {
+            Problem problem = problems.get(i);
+            numTestCases += problem.getTestCases().size();
+
+            ProblemContainer problemContainer = new ProblemContainer();
+            problemContainer.setProblem(problem);
```

Comment:
What will happen if a problem is deleted in the middle of a game? I believe the problem will still exist in the game object since it's in the backend, but what will happen here?

Reply:
This is a good point! You're right - I believe the problem will still exist in the game object, but it will not exist in the database. The following error thus exists.

**Problem Deleted During Game**
<img width="569" alt="Problem Deleted During Game" src="https://user-images.githubusercontent.com/7987279/126513523-1d227368-d85a-4877-9290-c9da0d0f1dcd.png">

I therefore want to make the change to check if the problem does not exist in the database, then I want to set the problem within the `ProblemContainer` as null. I'm adding this, and I'm also going to do this right before I save the game report.

There is still a chance, I believe, that while looping through the problems, one of the problems is deleted after it is already visited. I'm not sure what to do about conflicts like this - I think that relates to the issue we've had elsewhere of simultaneous SQL updates.

Reply:
I will note that after I implemented this change, it worked, but still produced the below error in the Terminal; I'm not sure whether this is something that I should be concerned about or not.

**Terminal Problem Not Found Error**
<img width="568" alt="Terminal Problem Not Found Error" src="https://user-images.githubusercontent.com/7987279/126520706-c1eb58ee-933c-43c3-8b6d-89d0a19ed174.png">

Reply:
Does this error still happen? If so, I wonder if it has to do with one of the timer tasks not canceling properly? If it doesn't affect anything then should hopefully be fine, assuming that it only shows up in the edge case of deleting a problem